core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java

```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
 
     @Override
     public RegisteredService findServiceBy(final long id) {
-        val result = this.services.get(id);
+        val result = this.services.get(id, k -> this.serviceRegistry.findServiceById(id));
         return validateRegisteredService(result);
     }
 
+    @Override
+    public <T extends RegisteredService> T findServiceBy(final long id, final Class<T> clazz) {
+        val result = this.serviceRegistry.findServiceById(id, clazz);
+        if (result != null) {
+            services.get(result.getId(), k -> result);
+        }
+        return result;
+    }
```

Comment:
This doesn't look right. Should be:


```
result = services.get(result.getId(), k -> this.serviceRegistry.findServiceById(id, clazz));
```

Also, don't forget to check the result to make sure it's has the correct type based on `clazz`